ros2bag/ros2bag/reindexer/reindex_sqlite.py

```diff
@@ -0,0 +1,178 @@
+# Part 252.227-7013 or 7014 (Feb 2014).
+#
+# This notice must appear in all copies of this file and its derivatives.
+
+import pathlib
+import sqlite3
+import sys
+from typing import List, Literal, Optional, TypedDict
+
+from ros2bag.api import print_error
+
+from . import bag_metadata
+
+
+class TopicInfo(TypedDict):
+    topic_name: str
+    topic_type: str
+    topic_ser_fmt: str
+    topic_count: str
+    topic_qos: str
+
+
+class DBMetadata(TypedDict):
+    topic_metadata: List[TopicInfo]
+    min_time: int
+    max_time: int
+
+
+def get_metadata_from_connection(db_con: sqlite3.Connection) -> DBMetadata:
+    # Query the metadata
+    c = db_con.execute('SELECT name, type, serialization_format, COUNT(messages.id), '
+                       'MIN(messages.timestamp), MAX(messages.timestamp), offered_qos_profiles '
+                       'FROM messages JOIN topics on topics.id = messages.topic_id '
+                       'GROUP BY topics.name;')
+
+    rows = c.fetchall()
+
+    # Set up initial values
+    topics: List[TopicInfo] = []
+    min_time: int = sys.maxsize
+    max_time: int = 0
+
+    # Aggregate metadata
+    for row in rows:
+        topics.append(TopicInfo(
+            topic_name=row[0],
+            topic_type=row[1],
+            topic_ser_fmt=row[2],
+            topic_count=row[3],
+            topic_qos=row[6]))
+        if row[4] < min_time:
+            min_time = row[4]
+        if row[5] > max_time:
+            max_time = row[5]
+
+    return {'topic_metadata': topics, 'min_time': min_time, 'max_time': max_time}
+
+
+def get_metadata_file_compressed(db_file: pathlib.Path) -> DBMetadata:
+    try:
+        import zstandard as zstd
+    except ImportError:
+        raise ImportError(print_error(
+                            'The "zstandard" library is required to reindex compressed bags. '
+                            'Install using "pip3 install zstandard" and try again.'))
+
+    # Decompress database
+    compressed_db_bytes = db_file.read_bytes()
+    dctx = zstd.ZstdDecompressor()
+    decompressed_db = dctx.decompress(compressed_db_bytes)
+
+    # Temporarily save decompressed database to disk
+    decompressed_path = pathlib.Path(db_file.parent) / db_file.stem
+    decompressed_path.write_bytes(decompressed_db)
+    with sqlite3.connect(decompressed_path) as db_con:
+        return_val = get_metadata_from_connection(db_con)
+
+    # Delete temporary database, shm, and wal files
+    decompressed_path.unlink()
+    decompressed_shm = pathlib.Path(decompressed_path.as_posix() + '-shm')
+    decompressed_shm.unlink(True)
+    decompressed_wal = pathlib.Path(decompressed_path.as_posix() + '-wal')
+    decompressed_wal.unlink(True)
+
+    return return_val
+
+
+def get_metadata(db_file: pathlib.Path,
+                 compression_fmt: Literal['', 'zstd'],
+                 compression_mode: Literal['', 'none', 'file', 'message']) -> DBMetadata:
+    # Handle compression
+    if compression_fmt != '':
```

Comment:
```suggestion
    if compression_fmt:
```

handles both `None` and `''`. I would recommend using `None` rather than empty-string for optional values.

Also - this block would be much more readable as

```
if compression_fmt in ['none', 'message']:
  # open normally
elif compression_fmt == 'file':
  # open compressed
else:
  raise ValueError('unknown compression mode')
```

The "by process of elimintation !=" pattern is a lot less readable, and includes some duplicated lines.